docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+* Exit flow (Update, Re-Check). Update: unconditionally; Re-Check: only if App says TX is now invalid
+  * Transactions in higher-priority lanes are processed (updated) first.
+  * Depends on how the multiple TX lists are implemented (see discussion above)
+  * Update contains Re-check at the end
+  * Remember Update is done while holding the mempool lock (no new CheckTx calls... they have to wait)
+  * Changes:
+    * We could update the different TX lists in parallel
+    * `Update` method (implementing `Update` method in `Mempool` interface) is implemented by `CListMempool` so
+      * If we go for several `CListMempool`s, then we'll have to call `Update` on each of them with all the TXs every time
+        * or move `Update` method out of `CListMempool` (where?)
+      * Also, the way we will do `ReCheck` would be complicated if we had N `Update` calls
+        * We will do `ReCheck` lane-by-lane, in FIFO order within a lane
+      * So, this seems to make us lean toward: one `CListMempool` containing N `txs` lists
+* Each lane informs consensus when there are txs available.
+  *  `TxsAvailable` is in our laundry list
+
+ADR: Core part of detailed design. Bullets will likely become subsections
+
+### Who defines lanes and priorities
+
+The list of lanes and their priorities: are consensus parameters? are defined by the app?
+* Discussion on _where_ the lane configuration is set up. Three options
+  1. `config.toml` / `app.toml`
+  2. ConsensusParams
+  3. Hardcoded in the application. Info passed to CometBFT during handshake
+* Outcome of discussion
+  * 1. vs. 2.
+    * It does not make sense for different nodes to have different lane configuration,
+      so definitely we don't want 1
+  * 2. vs. 3.
+    * If we can change lane info via `ConsensusParams`
+      * CometBFT's mempool needs logic for changing lanes dynamically (complex, not really appealing for MVP)
+      * The process of updating lane info very complex and cumbersome:
+        * to update lanes you'd need to pass _two_ governance proposals
+          * upgrade the app, because the lane classification logic (so, the app's code) needs to know the lane config beforehand
+          * then upgrade the lanes via `ConsensusParams`
+        * also, not clear in which order: community should be careful not to break performance between the passing of both proposals
+        * the `gov` module may allow the two things to be shipped in the same gov proposal,
+          but, if we need to do it that way, what's the point in having lanes in `ConsensusParams` ?
+  * Conclusion: 3.
+    * lane info is "hardcoded" in the app's logic
+    * lane info is passed to CometBFT during handshake
+  * Advantages:
+    * Simple lane update cycle: one software update gov proposal
+    * CometBFT doesn't need to deal with dynamic lane changes: it just needs to set up lanes when starting up (whether afresh, or recovery)
```

Comment:
Would updating the app and the lanes render two nodes incommunicable?

I think we should consider the case where two nodes don't fully share their lanes setup.

Reply:
In the current design, updating lane info requires a coordinated update. That simplifies this design a lot. We need to check with users whether this fits their use cases (mainly, that the don't want to change lanes every day). But if have to drop the "updating lane info requires a coordinated update" requirement, the design will become _way_ more complex. As in every design exercise, it's a tradeoff we must carefully consider.